### ISSUE TITLE
[Epic] Expose error stream via SQL ("bad data"/"programmable errors")

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#22430]
## Background

In Materialize, every dataflow collection has an associated error stream, representing errors that occurred while computing that collection. For example, consider a simple table with columns `(a int, b int)`, and the following query:

```
CREATE MATERIALIZED VIEW v AS SELECT a / b FROM t
```

If `(1, 0)` is inserted into `t`, the materialized view `v`'s error stream will contain a single division by zero error. If `(2, 0)` is inserted into `t`, then the materialized view `v`'s error stream will contain _two_ division by zero errors. If both of those tuples are later removed from `v`, the error stream will contain no errors.

To match the behavior of other SQL databases, Materialize today will only let you interact with the error stream in one very specific way: during queries of the collection. A query that reads data from `v` (e.g., `SELECT * FROM v`) will return an error if there are any errors in the error stream. If there is more than one error, it picks one at random to return.

This is not very ergonomic. You can't see how _many_ errors there are; you can't see more than one error at once; you can't ask the simple question of "are there any errors in this collection?" without also load the data in the collection if there are no errors. If you want to fix the errors, you have to fix the one error you see in order to move on to the next error.

## Proposal

At its core, the proposal is simple: make the error stream accessible to SQL. This lets users transform the error stream using the SQL syntax they're already familiar with to answer whatever questions they have about a collection's errors ("are there any?", "how many are there?", "show me the first 20 errors").

### Alternative 1

One option is to introduce a name for the error stream, like `<object>$errors`, which implicitly exists for every collection and contains all of the collection's errors:

```
> SELECT * FROM v$errors
code | name | description | hint
-----|------|-------------|----
```

We could also consider introducing an "oks" stream, which works like the base collection, except querying returns all non-errors in the collection, even if there are errors in the error stream.

One cool thing about this proposal is that you could extend it to intermediate relations in a query:

```
WITH t AS (
    SELECT a / b FROM whatever
)
SELECT * FROM t$errors
```

### Alternative 1.5

Like alternative 1, but introduce some special function-like syntax, e.g.:

```sql
SELECT * FROM ERRORS(v), PROGRESS(v), OKS(v);

-- It's valid to put additional relational expressions inside:
SELECT * FROM ERRORS(SELECT * FROM );
```

This seems tricky to parse, but not impossibly hard, and avoids some difficult questions about "what do you do if someone creates a view literally named `v$errors`?" that come up with alternative 1. 

Note these can't be normal functions because normal functions can't take tables as input.

### Alternative 2

We could extend the `CREATE [MATERIALIZED] VIEW` syntax to allow specifying a name for the error stream. This is consistent with what we do for progress collections for sources.

```
CREATE [MATERIALIZED] VIEW v WITH (
    EXPOSE ERRORS AS v_errors
) AS ...
```

The downside is that this would only allow users to query error streams for named objects in the catalog; there'd be no way to query error streams for intermediate relations in a one-off query.

### "Oks" stream

An important of programmable errors is the ability to _ignore_ errors. We should also expose the underlying "oks" stream, which consists of all valid data without any errors. Imagine:

```
-- Alternative 1
> SELECT * FROM v$oks

-- Alternative 2
CREATE [MATERIALIZED] VIEW v WITH (
    EXPOSE ERRORS AS v_errors,
    EXPOSE OKS AS v_oks
) AS ...
SELECT * FROM v_oks
```

where the `$oks` suffix means you're only interested in looking at the valid data in `v`—you _don't_ want the query to fail if `v` contains one or more errors.

We should workshop better terms for this than "ok"/"oks", as that terminology is only familiar to those who know Rust's `Result` type. Unfortunately I don't think there's any prior art within SQL to guide us.

## See also

* https://github.com/MaterializeInc/materialize/issues/17668

cc @bkirwi @umanwizard @frankmcsherry 

## Action log

* *13 November 2023.* Added "Alternative 1.5" section.
* *11 November 2023.* Added "Oks stream" section.
* *17 October 2023.* Initial issue filed.

**Comment by teskje** [comment]
> You can't see how many errors there are; you can't see more than one error at once; you can't ask the simple question of "are there any errors in this collection?" without also load the data in the collection if there are no errors.

FWIW, it is now possible to see how many errors there are (and therefore also if there are any errors) in each collection by querying `mz_compute_error_counts`. Not trying to imply that this is particularly ergonomic of course! And there is still no way to see _what_ errors there are.

**Comment by benesch** [comment]
Oho! Good to know. TIL.

**Comment by jkosh44** [comment]
What about a regular function on the column that returns the error if there is one and `NULL` otherwise. Something like:

```SQL
CREATE TABLE t (a INT, b INT);

CREATE [MATERIALIZED] VIEW v (av, bv) AS SELECT 1/a, 1/(b-1) FROM t;

INSERT INTO t VALUES (0, 0), (1, 1);

SELECT ERROR(av), ERROR(bv) FROM v;
---
'Div by 0'    |    NULL
NULL          |    'Div by 0'
```

**Comment by umanwizard** [comment]
I think there can be errors that are not associated with any particular column.

E.g.

```
CREATE MATERIALIZED VIEW v AS SELECT * FROM t WHERE a/b > 42;
```

**Comment by frankmcsherry** [comment]
Drive-by thoughts: The "Oks Stream" proposal is the least scary, in that all of the others treat errors as data, and to date there are no specific semantics around errors. No specific reason to believe they'll be stable across replicas, that they satisfy any properties we commonly assume about data, etc. You could imagine loading them out of persist collections an data, because at least there something specific has been written down (perhaps inexplicably).

**Comment by jkosh44** [comment]
> I think there can be errors that are not associated with any particular column.
> 
> E.g.
> 
> ```
> CREATE MATERIALIZED VIEW v AS SELECT * FROM t WHERE a/b > 42;
> ```
> 
> can divide by 0 in the where clause.

Oh interesting, so for the current proposals, if I have some view defined like

```SQL
CREATE MATERIALIZED VIEW v AS SELECT 1/a FROM t WHERE 1/b > 42
```

and I add the following row

```SQL
INSERT INTO t VALUES (0, 0);
```

would that produce a single error in the error stream or two errors in the error stream?

**Comment by umanwizard** [comment]
My hunch is only 1, but I wouldn't rely on that, because we never really defined meaningful semantics for errors (unless that has happened since I left MZ, which I doubt).

**Comment by sjwiesman** [comment]
The specific concern that prompted re-exploring this issue is handling malformed data in sources. And I have a potentially 🌶️ take that malformed data on Kafka is not an error but a case that the system should gracefully handle. 

As a data warehouse, Materialize collects facts generated by other systems. The data warehouse's job is to collect, clean, and process this data for derived data products. The warehouse must be able to handle whatever sludge is thrown at it. For example, consider Snowflake, which supports error handling when loading files[1]. Or various stream processors that use the dead letter queue pattern[2]. Contrast this with an OLTP database where correctness properties can be asserted when inserting records into a table. 

Let's break down the various sources for getting data into Materialize today: 

Webhooks: Webhook sources do a great job at enforcing invariants today. Materialize will reject requests whose body does not meet the specified format. If the system expects JSON, only JSON is ever written down into persist. 
Postgres / MySQL: Much like how we rely on these databases to provide an ordering for their data that our system respects, we also rely on them to provide correctness of the format. No one is concerned with Postgres writing random bytes into the publication in the happy case.
Kafka: Kafka is the weird one. While the best practice is to use a schema registry or other enforcement mechanism, Kafka topics only see bytes, and there is no reason to expect users not to publish malformed data. Today, the system does not have any mechanism for handling malformed data. If you are using upsert the source is wedged until a new record is written with the same key, which may be tricky if the Mateiralize team does not own the topic. And if the sources is append only source there is no solution. 

Issues such as divide by zero represent a business logic error. Errors arising from reduce that are likely bugs in our system. Dealing with bad data in Kafka is an inevitable part of working with the system. Although I appreciate the proposed solution outlined in this issue, I am concerned that it may be idealistic without being practical to users needs.

[1] https://docs.snowflake.com/en/sql-reference/sql/copy-into-table#copy-options-copyoptions
[2] https://www.confluent.io/blog/kafka-connect-deep-dive-error-handling-dead-letter-queues/

**Comment by benesch** [comment]
> > You can't see how _many_ errors there are;
> 
> If the hunch is true, it seems like the proposal doesn't fully fix this issue, but it does move in the right direction. Instead of seeing one error per collection, we'd see one error per row.

Nah, I'd call the proposal a full fix for the issue! What I meant there about error counts is something specific: we have an errs stream that contains any number of errors, but there is no way today for a user to see the count of errors in the errs stream.

The design of the oks and errs streams are described here: https://github.com/MaterializeInc/materialize/blob/d2061d4edafd03d8be5381698a40a2e191503409/src/compute/src/render.rs#L12-L101

> all of the others treat errors as data, and to date there are no specific semantics around errors

Yes, the customer/product ask is to be able to monitor and alert on errors. I think the best way to satisfy this ask is to allow users to treat errors as data.

> No specific reason to believe they'll be stable across replicas, that they satisfy any properties we commonly assume about data, etc.

To flip this around: is there a specific reason to believe that errors will _not_ be stable across replicas, or violate properties that we commonly assume about data? 

I _believe_ we already enforce an important invariant about errors across replicas. For a given LIR plan, we expect all replicas to agree about what error to return (if any). If different replicas returned different errors, or if one replica returned an error while another replica returned data, we'd consider that a bug today, no?

> Although I appreciate the proposed solution outlined in this issue, I am concerned that it may be idealistic without being practical to users needs.

Can you say more about what you find idealistic about the proposal? To me, it is a relatively direct mapping of the idea of a DLQ onto SQL, with some care taken to ensure backwards compatibility with our current semantics. The `ERRORS(x)` collection is the DLQ, and the `OKS(x)` collection is the "I promise I'm checking the DLQ elsewhere, just show me the data" collection. If we can come up with short but clear names for `ERRORS` and `OKS`, it seems quite ergonomic to use, too.

**Comment by antiguru** [comment]
> To flip this around: is there a specific reason to believe that errors will not be stable across replicas, or violate properties that we commonly assume about data?

I don't think this is true: In the partitioned compute client we collect responses per part, which corresponds to either a timely worker or a process. We then return the first error in the order of parts. If we have two replicas in a cluster of different size, the partitioning of data can be different and thus the worker where the error occurs might change, potentially yielding a different error message. If we were to collect and sort the errors, we might be able to mitigate the problem.

Similarly for subscribes, where we return per worker the first error we encounter in a batch of updates.

**Comment by frankmcsherry** [comment]
> would that produce a single error in the error stream or two errors in the error stream?

It would almost certainly be a single error, at the moment, because predicate evaluation (and most evaluation) that results in an error does not result in data. So, erroring evaluations consume the data as they result in the error.

On the other hand, some "smart" optimizer could change the `WHERE` clause into a semijoin, and push down the `1/a` selection, and now each of the inputs to the semijoin would produce and error and there would be two. Generally, no semantics around errors that I'm aware of, other than that error-free computations should be error-free.

**Comment by frankmcsherry** [comment]
> And I have a potentially 🌶️ take that malformed data on Kafka is not an error but a case that the system should gracefully handle.

I .. think I agree with this. At least, in trying to mentally unpick the errors that are going on here, in analogy to Rust there are fatal errors (`assert`) and there are recoverable errors (`Result`). 

1. Source ingestion from a system that does not ensure data match specified schemas (e.g. Kafka) feel like recoverable errors, and ideally we could have the pressure release valve of a DLQ at that moment, if not something more general. 
2. Once in MZ, i.e. once we've asserted that we've introduced certain structure (right type, primary keys respected, no negative accumulations), errors feel a fair bit more fatal. In the case of ingestion, we are computing a specific answer on a modification of your input data. In the case of "your subquery returned too many values" we are computing .. no specific thing on your data any more; we have no idea how wrong it is, only that something went wrong enough that we can no longer clearly communicate that wrongness in the data itself.
3. There are maybe things in between, like "your source data are `text`, and you expected to cast them to JSON and find a specific field", which .. idk. The whole Big Data vs SQL argument was some amount about this, and it feels a bit like you might want a similar architectural boundary: when do we shift from "unstructured chaos" to "validated and expected to work".

Strong feels that if you want to treat errors as data, you should describe the errors as data also. Right now they are "described" by the dataflows that are built in rendering and the scalar function implementations that may or may not return errors. We could choose to be more specific about them, much like Rust would ask you to be quite specific about what you plan to do with the `Result` type, other than `longjmp` back to a try/catch header, which feels closer to way we currently treat errors and what we might back ourselves into by publishing error collections.

**Comment by benesch** [comment]
> In the case of "your subquery returned too many values" we are computing .. no specific thing on your data any more; we have no idea how wrong it is, only that something went wrong enough that we can no longer clearly communicate that wrongness in the data itself.

FWIW, sinks today ignore the errors collection and just keep sinking whatever happens to be the oks collection: https://github.com/MaterializeInc/materialize/issues/17668. It's a fairly large glitch in our correctness story, but also light evidence that the data in the oks collection is "right enough" to be useful even in the face of errors.

> Strong feels that if you want to treat errors as data, you should describe the errors as data also. Right now they are "described" by the dataflows that are built in rendering and the scalar function implementations that may or may not return errors. We could choose to be more specific about them, much like Rust would ask you to be quite specific about what you plan to do with the `Result` type, other than `longjmp` back to a try/catch header, which feels closer to way we currently treat errors and what we might back ourselves into by publishing error collections.

How would you graft `Result` types onto SQL in a backwards compatible manner? It seems like you'd need to introduce a `Result` type and a new function call syntax that returns errors as `Results` rather than throwing them into the errs stream. That's a pretty substantial change that would make Materialize SQL much less interoperable with other flavors of SQL.

**Comment by teskje** [comment]
> I don't think this is true: In the partitioned compute client we collect responses per part, which corresponds to either a timely worker or a process. We then return the first error in the order of parts. If we have two replicas in a cluster of different size, the partitioning of data can be different and thus the worker where the error occurs might change, potentially yielding a different error message. If we were to collect and sort the errors, we might be able to mitigate the problem.

I don't think you even need different replica sizes, just workers in different replicas returning peek responses in different orders. But this inconsistency between replicas only occurs during peek processing and is caused by the fact that peeks only return a single error randomly. For the question of whether or not a `SELECT * FROM v$errors` syntax could be sound we instead care about whether or not the error streams (i.e. the contents of the error arrangements of `v`) of different replicas are the same, no? Do we have any reason to believe they are not?

If we do have reason to believe replicas can have different error streams, perhaps that's something we should try to fix. At the very least this would cause issues for materialized views where the self-correcting persist sinks would fight to append their version of the error updates, producing a bunch of churn on persist.

**Comment by benesch** [comment]
> I don't think you even need different replica sizes, just workers in different replicas returning peek responses in different orders. But this inconsistency between replicas only occurs during peek processing and is caused by the fact that peeks only return a single error randomly.

Eep, yeah, seems less than ideal. Big +1 from me on waiting for all workers to respond and then sorting the errors to ensure we have a stable response.

> If we do have reason to believe replicas can have different error streams, perhaps that's something we should try to fix. At the very least this would cause issues for materialized views where the self-correcting persist sinks would fight to append their version of the error updates, producing a bunch of churn on persist.

Agree!

**Comment by teskje** [comment]
> I don't think you even need different replica sizes, just workers in different replicas returning peek responses in different orders. 

Turns out this is wrong. The `PartitionedComputeState` already waits for all responses before emitting an error upstream, so it will always send the error from the worker with the smallest ID, so things are fine if all replicas have the same number of workers.

> Big +1 from me on waiting for all workers to respond and then sorting the errors to ensure we have a stable response.

Turns out it's not that simple. We can easily sort and return, say, the "smallest" error in the `PartitionedComputeState`, but then we also need to make sure that each worker always returns the "smallest" error from its error arrangement, which I think involves converting all errors to strings and then sorting them (given that the arrangement sorting doesn't need to match the string sorting). [Tracking issue.](https://github.com/MaterializeInc/materialize/issues/27223)

**Comment by bosconi** [comment]
@benesch awaiting your design notes.

**Comment by benesch** [comment]
I can commit to getting a design down in writing by start of business on Monday.

**Comment by umanwizard** [comment]
Closed accidentally? Looks like the linked PR is only for upsert sources.

**Comment by benesch** [comment]
Agreed! Reopening to track that `IGNORE ERRORS` and `ONLY ERRORS` are still outstanding. We may ultimately not want to pursue those options, but nonetheless seems worth keeping the issue around.

**Comment by bosconi** [comment]
Per discussion with @benesch, moving to backlog. We only needed this for Kafka upsert source right now.

https://github.com/MaterializeInc/materialize/issues/28127 tracks the MVP work, including potential followup work such as `IGNORE ERRORS` and `ONLY ERRORS`.
